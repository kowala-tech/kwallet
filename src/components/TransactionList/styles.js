const styles = (theme) => ({
	card: {
		marginBottom: theme.spacing.unit * 2,
		paddingBottom: theme.spacing.unit
	},
	stickyHeader: {
		backgroundColor: theme.palette.background.paper,
	},
	list: {
		width: "100%",
		backgroundColor: theme.palette.background.paper,
		position: "relative",
		overflow: "auto"
	},
	avatar: {
		backgroundColor: theme.palette.primary.main
	}
});

export default styles;
