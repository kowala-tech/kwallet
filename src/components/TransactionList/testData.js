import moment from "moment";

const buildTransactions = (n=30) => {
	let arr = [];
	for (let i = 0; i < n; i++) {
		arr.push(
			{
				to: "32Be343B94f860124dC4fEe278FDCBD38C102D88",
				from: null,
				nativeAmount: 1000000000000000000,
				fees: 0.01,
				id: "0x6c90256274c87ed74783d1d049304110377005b016af3661c24a71321cf0c0ec",
				timestamp: moment().subtract(Math.floor(Math.random() * Math.floor(14)), "day").format()
			}
		);
	}
	return arr;
};

const transactions = buildTransactions();

export default transactions;
