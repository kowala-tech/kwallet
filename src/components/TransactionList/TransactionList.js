import React from "react";
import PropTypes from "prop-types";
// Material UI Imports
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import ListSubheader from "@material-ui/core/ListSubheader";
import Divider from "@material-ui/core/Divider";
import Typography from "@material-ui/core/Typography";
import Avatar from "@material-ui/core/Avatar";
import Card from "@material-ui/core/Card";
import CardHeader from "@material-ui/core/CardHeader";
import CardContent from "@material-ui/core/CardContent";
import CardActions from "@material-ui/core/CardActions";
import Grow from "@material-ui/core/Grow";
import IconButton from "@material-ui/core/IconButton";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import ListIcon from "@material-ui/icons/List";
// Component Related Imports
import ListItem from "./components/ListItem";
import Progress from "../Progress";
import styles from "./styles";
import { unixToHumanDate } from "../../transforms/time";

const groupBy = (array) => {
	return array.reduce(function(groups, item) {
		const timestamp = unixToHumanDate(item["date"]);
		groups[timestamp] = groups[timestamp] || [];
		groups[timestamp].push(item);
		return groups;
	}, {});
};

function TransactionList( { classes, loading, transactions, toggleTransactionDetailsDialogFn, depositButton } ) {
	if (loading) {
		return (<Progress/>);
	}

	const groupedTransactions = groupBy(transactions);

	return (
		<Grow in>
			<Card className={classes.card}>
				<CardHeader
					avatar={
						<Avatar className={classes.avatar}>
							<ListIcon />
						</Avatar>
					}
					action={
						<IconButton disabled>
							<MoreVertIcon />
						</IconButton>
					}
					title={
						<Typography variant="subheading">
							Transactions
						</Typography>
					}
				/>
				{ Object.keys(groupedTransactions).length === 0 && (
					<div>
						<CardContent>
							<Typography variant="body1">It looks like you don't have any kUSD yet! To get started, deposit some kUSD into this wallet.</Typography>
						</CardContent>
						<CardActions
							className={classes.actions}
						>
							{depositButton}
						</CardActions>
					</div>
				)}
				<div>
					<List
						dense
						disablePadding
						className={classes.list}
					>
						{
							Object.keys(groupedTransactions).sort().reverse().map( (date, index) => {
								return(
									<div key={index}>
										<ListSubheader className={classes.stickyHeader}>
											{date}
											<Divider/>
										</ListSubheader>
										{(
											groupedTransactions[date].map( (transaction, index) => {
												return(
													<ListItem
														button
														key={index}
														transaction={transaction}
														onClick={toggleTransactionDetailsDialogFn}
													/>
												);
											})
										)}
									</div>
								);
							})
						}
					</List>
				</div>
			</Card>
		</Grow>
	);
}

TransactionList.propTypes = {
	transactions: PropTypes.array.isRequired,
	loading: PropTypes.bool.isRequired,
	classes: PropTypes.object.isRequired,
	toggleTransactionDetailsDialogFn: PropTypes.func,
	depositButton: PropTypes.object
};

export default withStyles(styles)(TransactionList);
