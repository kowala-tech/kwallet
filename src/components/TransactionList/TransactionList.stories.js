import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
// Story related component imports
import TransactionList from "./TransactionList";
import transactions from "./testData";

storiesOf("Transaction List", module)
	.add("loading", () =>
		(<TransactionList
			loading={true}
			transactions={[]}
			toggleTransactionDetailsDialogFn={action("transaction-details-click")}
		/>)
	)
	.add("without transactions", () =>
		(<TransactionList
			loading={false}
			transactions={[]}
			toggleTransactionDetailsDialogFn={action("transaction-details-click")}
		/>)
	)
	.add("with transactions", () =>
		(<TransactionList
			loading={false}
			transactions={transactions}
			toggleTransactionDetailsDialogFn={action("transaction-details-click")}
		/>)
	);
