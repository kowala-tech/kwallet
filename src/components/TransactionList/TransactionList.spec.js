import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import moment from "moment";
// Story related component imports
import TransactionList from "./TransactionList";

describe("TransactionList", () => {

	it("it should match the snapshot", () => {
		const transactions = [
			{
				to: "32Be343B94f860124dC4fEe278FDCBD38C102D88",
				from: null,
				nativeAmount: 1000000000000000000,
				fees: 0.01,
				id: "0x6c90256274c87ed74783d1d049304110377005b016af3661c24a71321cf0c0ec",
				timestamp: moment("2018-01-01T00:00:00-00:00").utc()
			}
		];
		const component = toJson(
			shallow(
				<TransactionList
					loading={false}
					transactions={transactions}
					toggleTransactionDetailsDialogFn={jest.fn()}
				/>
			)
		);
		expect(component).toMatchSnapshot();
	});

});
