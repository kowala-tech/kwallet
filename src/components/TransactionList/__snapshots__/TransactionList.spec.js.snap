// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList it should match the snapshot 1`] = `
<TransactionList
  classes={
    Object {
      "avatar": "TransactionList-avatar-4",
      "card": "TransactionList-card-1",
      "list": "TransactionList-list-3",
      "stickyHeader": "TransactionList-stickyHeader-2",
    }
  }
  loading={false}
  toggleTransactionDetailsDialogFn={[MockFunction]}
  transactions={
    Array [
      Object {
        "fees": 0.01,
        "from": null,
        "id": "0x6c90256274c87ed74783d1d049304110377005b016af3661c24a71321cf0c0ec",
        "nativeAmount": 1000000000000000000,
        "timestamp": "2018-01-01T00:00:00.000Z",
        "to": "32Be343B94f860124dC4fEe278FDCBD38C102D88",
      },
    ]
  }
/>
`;
